node/src/components/block_validator.rs

```diff
@@ -59,114 +68,114 @@ pub(crate) struct BlockValidationState<T> {
+    ) -> Effects<Event<T, I>>
+    where
+        REv: From<Event<T, I>>
+            + From<BlockValidationRequest<T, I>>
+            + From<StorageRequest>
+            + From<FetcherRequest<I, Deploy>>
+            + Send,
+    {
+        let mut effects = Effects::new();
         match event {
             Event::Request(BlockValidationRequest {
                 block,
                 sender,
                 responder,
                 block_timestamp,
             }) => {
-                if block.deploys().is_empty() {
-                    // If there are no deploys, return early.
-                    let mut effects = Effects::new();
-                    effects.extend(responder.respond((true, block)).ignore());
-                    return effects;
-                }
-                // No matter the current state, we will request the deploys inside this protoblock
-                // for now. Duplicate requests must still be answered, but are
-                // de-duplicated by the fetcher.
-                let effects = block
+                let block_deploys = block
                     .deploys()
                     .iter()
-                    .flat_map(|deploy_hash| {
-                        // For every request, increase the number of in-flight...
-                        self.in_flight.inc(*deploy_hash);
-
-                        // ...then request it.
-                        let dh_found = **deploy_hash;
-                        let dh_not_found = **deploy_hash;
-                        effect_builder
-                            .fetch_deploy(**deploy_hash, sender.clone())
-                            .map_or_else(
-                                move |result: FetchResult<Deploy>| match result {
-                                    FetchResult::FromStorage(deploy)
-                                    | FetchResult::FromPeer(deploy, _) => {
-                                        if deploy.header().timestamp() > block_timestamp {
-                                            Event::DeployMissing(dh_found)
-                                        } else {
-                                            Event::DeployFound(dh_found)
-                                        }
-                                    }
-                                },
-                                move || Event::DeployMissing(dh_not_found),
-                            )
-                    })
-                    .collect();
+                    .map(|deploy_hash| **deploy_hash)
+                    .collect::<HashSet<_>>();
+                if block_deploys.is_empty() {
+                    // If there are no deploys, return early.
+                    return responder.respond((true, block)).ignore();
+                }
 
                 // TODO: Clean this up to use `or_insert_with_key` once
                 // https://github.com/rust-lang/rust/issues/71024 is stabilized.
                 match self.validation_states.entry(block) {
```

Comment:
How long does the block validator store the information about the block's validation state? Once a block is validated, is the entry removed from `validation_states`?

Reply:
It is only dropped if a block _fails_ validation, and that behaviour existed before my changeset.